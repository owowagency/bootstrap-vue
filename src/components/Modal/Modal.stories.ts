import Btn from '@/components/Button';
import Modal from '.';
import {sizes} from '@/composables/useSize';

export default {
    title: 'Components/Modal',
    component: Modal,
    argTypes: {
        hideCloseBtn: {
            control: {type: 'boolean'},
        },
        modalCentered: {
            control: {type: 'boolean'},
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
    },
};

const template = (args) => ({
    components: {Btn, Modal},
    setup: () => ({args}),
    template: `
        <Modal
            class="d-block"
            title="Modal title"
            text="Modal body text goes here."
            v-bind="args"
        >
            <template #footer>
                <Btn variant="secondary">
                    Close
                </Btn>

                <Btn>
                    Save changes
                </Btn>
            </template>
        </Modal>
    `,
});

export const Default = template.bind({});

export const Demo = (args) => ({
    components: {Btn, Modal},
    setup: () => ({args}),
    template: `
        <Btn
            data-bs-toggle="modal"
            data-bs-target="#modal"
        >
            Launch demo modal
        </Btn>

        <Modal
            id="modal"
            title="Modal title"
            text="Modal body text goes here."
            v-bind="args"
        >
            <template #footer>
                <Btn
                    data-bs-dismiss="modal"
                    variant="secondary"
                >
                    Close
                </Btn>

                <Btn>
                    Save changes
                </Btn>
            </template>
        </Modal>
    `,
});
