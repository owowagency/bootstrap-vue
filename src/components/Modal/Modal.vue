<template>
    <div
        :id="id"
        ref="modal"
        class="modal"
    >
        <div
            class="modal-dialog"
            :class="classes"
        >
            <div class="modal-content">
                <div
                    v-if="!!$slots.header || !!$slots.title || !!title || !hideCloseBtn"
                    :class="['modal-header', headerClass]"
                >
                    <slot name="header">
                        <h5
                            v-if="!!$slots.title || !!title"
                            class="modal-title"
                        >
                            <slot name="title">
                                {{ title }}
                            </slot>
                        </h5>

                        <button
                            v-if="!hideCloseBtn"
                            class="btn-close"
                            data-bs-dismiss="modal"
                            type="button"
                        />
                    </slot>
                </div>

                <div
                    v-if="$slots.default || body"
                    :class="{'modal-body': !noBody}"
                >
                    <slot>
                        {{ body }}
                    </slot>
                </div>

                <div
                    v-if="$slots.footer"
                    class="modal-footer"
                >
                    <slot name="footer" />
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import {computed, onBeforeUnmount, ref} from 'vue';
import useSize, {sizeProps} from '@/composables/useSize';
import {idProps} from '@/composables/useId';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';
import useClasses from '@/composables/useClasses';

const modalEvents = ['show', 'shown', 'hide', 'hidden', 'hidePrevented'] as const;

type ModalEvent = typeof modalEvents[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    body: {
        type: String,
        default: undefined,
    },
    headerClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    hideCloseBtn: {
        type: Boolean,
        default: false,
    },
    ...idProps,
    modalCentered: {
        type: Boolean,
        default: false,
    },
    noBody: {
        type: Boolean,
        default: false,
    },
    ...sizeProps,
    title: {
        type: String,
        default: undefined,
    },
});

const emit = defineEmits<{(event: ModalEvent): void}>();

const modal = ref<HTMLElement>();

useBootstrapEmits(
    modal,
    modalEvents,
    emit,
    'modal',
);

const {bsInstance: bsModal} = useBootstrapInstance(
    'Modal',
    modal,
);

const {classes} = useClasses(computed(() => [
    props.modalCentered ? 'modal-dialog-centered' : null,
    props.size ? useSize(props.size, 'modal-{0}').sizeClass.value : null,
]));

onBeforeUnmount(() => {
    if (bsModal.value) {
        bsModal.value.hide();
    }
});

defineExpose({bsModal});
</script>
