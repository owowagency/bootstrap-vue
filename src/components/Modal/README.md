# Modal

```vue
<Modal hideCloseBtn>
    <form>
        ...
    </form>

    <template #footer>
        <Button>Submit</Button>
    </template>
</Modal>
```

## Imported Props

| Prop name | Description                   | Type                                       | Values | Default |
| --------- | ----------------------------- | ------------------------------------------ | ------ | ------- |
| id        | The id used for the component | string                                     | -      | `uuid`  |
| size      | The size of the component     | [Size](../../composables/useSize) (string) | -      | 'md'    |

<!-- TODO Auto-generate based on type? -->

## Events

| Name          | Args | Description                                                                                                       |
| ------------- | ---- | ----------------------------------------------------------------------------------------------------------------- |
| show          |      | This event fires immediately when the `show` instance method is called                                            |
| shown         |      | This event is fired when the modal has been made visible to the user                                              |
| hide          |      | This event is fired immediately when the `hide` instance method has been called                                   |
| hidden        |      | This event is fired when the modal has finished being hidden from the user                                        |
| hidePrevented |      | This event is fired when the modal is shown, its backdrop is static and a click outside of the modal is performed |

## Props

| Prop name       | Description                                         | Type                  | Values | Default   |
| --------------- | --------------------------------------------------- | --------------------- | ------ | --------- |
| body            | The body text of the modal                          | string                | -      | undefined |
| headerClass     | The classes that will be forwarded to the header    | string\|array\|object | -      | undefined |
| hideCloseBtn    | Defines if the close button should be hidden        | boolean               | -      | false     |
| modalCentered   | Defines if the modal should be vertically centered  | boolean               | -      | false     |
| modalScrollable | Defines if the modal should scrollable              | boolean               | -      | false     |
| noBody          | Defines if the `modal-body` class should be omitted | boolean               | -      | false     |
| title           | The title text of the modal                         | string                | -      | undefined |

## Slots

| Name    | Description                            | Bindings |
| ------- | -------------------------------------- | -------- |
| header  | Displays the header of the modal       |          |
| title   | Displays the header title of the modal |          |
| default | Displays the body of the modal         |          |
| footer  | Displays the footer of the modal       |          |
