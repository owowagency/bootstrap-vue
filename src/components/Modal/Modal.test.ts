import Modal from '.';
import {shallowMount} from '@vue/test-utils';

jest.mock('bootstrap', () => ({
    Modal: {
        getOrCreateInstance: jest.fn(),
    },
}));

const textPropTest = (propName: string) => {
    it(`renders ${propName}`, () => {
        const text = `I am a ${propName}`;

        const wrapper = shallowMount(Modal, {
            props: {[propName]: text},
        });

        expect(wrapper.find(`.modal-${propName}`).text()).toBe(text);
    });
};

describe('template', () => {
    componentRenderTest(Modal, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(Modal, 'header');

    componentSlotRenderTest(Modal, 'title');

    componentSlotRenderTest(Modal);

    componentSlotRenderTest(Modal, 'footer');

    ['title', 'body'].forEach(textPropTest);

    it('does not render modal-title when not given', () => {
        // Don't give title prop or slot.
        const wrapper = shallowMount(Modal);

        expect(wrapper.find('.modal-title').exists()).toBe(false);

        // Assert that .btn-close is still rendered.
        expect(wrapper.find('.btn-close').exists()).toBe(true);
    });

    ['show', 'shown', 'hide', 'hidden', 'hidePrevented'].forEach((event: string) => {
        componentBootstrapEventTest(
            Modal,
            '.modal',
            event,
            'modal',
        );
    });

    it('renders given custom classes', () => {
        const wrapper = shallowMount(Modal, {
            props: {
                id: 'owow-modal',
                class: 'custom-class',
                headerClass: 'header-class',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});

describe('onMounted', () => {
    it('sets bsModal', async() => {
        const wrapper = await shallowMount(Modal);

        expect((await import('bootstrap')).Modal.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.modal, {});

        // TODO: Unable to assert bsModal value.
    });

    it('registers modal even listeners', async() => {
        // TODO: Unable to assert events.
    });
});

describe('hideCloseBtn', () => {
    it('hides close button', async() => {
        const wrapper = await shallowMount(
            Modal,
            {props: {hideCloseBtn: true}},
        );

        expect(wrapper.find('[data-bs-dismiss="modal"]').exists()).toBeFalsy();
    });
});

// TODO: Unable to assert dispose call on bsModal.
