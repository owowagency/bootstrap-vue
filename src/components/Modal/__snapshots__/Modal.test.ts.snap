// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div id=\\"dont-change-please\\" class=\\"modal\\">
  <div class=\\"modal-dialog modal-md\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <!--v-if--><button class=\\"btn-close\\" data-bs-dismiss=\\"modal\\" type=\\"button\\"></button>
      </div>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`template renders given custom classes 1`] = `
"<div id=\\"owow-modal\\" class=\\"modal custom-class\\">
  <div class=\\"modal-dialog modal-md\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header header-class\\">
        <!--v-if--><button class=\\"btn-close\\" data-bs-dismiss=\\"modal\\" type=\\"button\\"></button>
      </div>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>"
`;
